File: matlab/analyze_eddy_tracks.m
%% analyze eddy tracks
%
% Douglas Cahl, PhD 
% University of South Carolina 2023

clear

plt_individual = 0;  % separate plot for each eddy track
highrescoast = 1;
lw = 1;
min_steps = 12; % minimum track length to plot
radar_dt = 0.5; % timestep in hours

%%
name_dir = 'data/results/';     % directory of data
name_pre = 'data2_';    % name prefix, e.g., data2_1.mat , data2_2.mat , ...
tracksfile = [name_dir name_pre 'tracks'];
load(tracksfile)
data = load('data/data2.mat'); % lons lats file
lons = data.lons;
lats = data.lats;
clear data


%% setup
lon_min = double(floor(10*min(lons(:)))/10);
lon_max = double(ceil(10*max(lons(:)))/10);
lat_min = double(floor(10*min(lats(:)))/10);
lat_max = double(ceil(10*max(lats(:)))/10);

%%
disp(['max gap between eddy identification in timesteps = ' num2str(eddy_track_time_param)])% - 1)])
disp(['number of tracks = ' num2str(length(timegap))])
for i = 1:length(timegap)
    eddy_track_length(i) = length(direction{i});
    intensity(i) = mean(num_streams{i});
    t(i) = mean(Time{i});
    d = eig1{i} + eig2{i};
    diameter(i) = mean(d); % 2x average of major and minor axes
    maxdiameter(i) = max(d); % 2x max of major and minor axes
    mindiameter(i) = min(d); % 2x max of major and minor axes

    if eddy_track_length(i) > 1
        lon = lon_center{i};
        lat = lat_center{i};
        [x,y,UTMzone] = geog2utm_nodisp(lon,lat);
        
        dx = x(end)-x(1);
        dy = y(end)-y(1);
        dist1 = sqrt(dx^2+dy^2)*1000;
        dt = Time{i}(end)-Time{i}(1);
        dt = dt*24*3600;
        dist(i) = dist1;
        speed(i) = dist1/dt;
    else
        dist(i) = nan;
        speed(i) = nan;
    end

end

disp(['max eddy track length in timesteps = ' num2str(max(eddy_track_length))])


%% tracks by length plots

edl = eddy_track_length;


[edl,b] = sort(edl,'descend');
if ~plt_individual
    figure('color','w');
    m_proj('Mercator','lon',[lon_min lon_max],'lat',[lat_min lat_max])
    
    if highrescoast
        m_gshhs_h('patch',[0.8 0.8 0.8]); % coastline
    else
        m_gshhs_l('patch',[0.8 0.8 0.8]); % coastline
    end
    levels = [-10 -25 -50 -100 -250 -500 -1000 -2500];
    m_etopo2('contour',levels,'edgecolor',0.5*[1 1 1],...
        'linewidth',0.5,'linestyle','-','ShowText','off');
    m_grid(); 
    hold on
end

counter = 0;
w3 = [];
for i = 1:length(edl)
    j = b(i);
    
    if eddy_track_length(j) < min_steps 
        continue
    end
    counter = counter + 1;
    
    w1 = mean(omega{j},'omitnan');
    w2 = w1*3600*24/360;
    w3(i) = w2;
    
    lon = lon_center{j};
    lat = lat_center{j};
    
    if plt_individual
        figure('color','w');
        m_proj('Mercator','lon',[lon_min lon_max],'lat',[lat_min lat_max])
        if highrescoast
            m_gshhs_h('patch',[0.8 0.8 0.8]); % coastline
        else
            m_gshhs_l('patch',[0.8 0.8 0.8]); % coastline
        end
        levels = [-10 -25 -50 -100 -250 -500 -1000 -2500];
        m_etopo2('contour',levels,'edgecolor',0.5*[1 1 1],...
            'linewidth',0.5,'linestyle','-','ShowText','off');
        m_grid(); 
        hold on
    end
    
    if w1 > 0
        m_plot(lon,lat,'-b.') % CCW (Northern: cyclonic/upwelling ; Southern, anticyclonic/downwelling)
    else
        m_plot(lon,lat,'-r.') % CW (Northern: anticyclonic/downwelling ; Southern, cyclonic/upwelling)
    end
    
    m_plot(lon(1),lat(1),'o','LineWidth',2,'color',[0 .8 0])
    m_plot(lon(end),lat(end),'kx','LineWidth',2)
    if i == 1 && ~plt_individual
        title(['longest track = ' num2str(eddy_track_length(j)) ' timesteps  ;   ' ...
            num2str(round(w2,1)) ' rot/day'])
    elseif plt_individual
        title(['track lenth = ' num2str(eddy_track_length(j)) ' timesteps  ;   ' ...
            num2str(round(w2,1)) ' rot/day'])
    end
end



%% histogram plots
ed1 = eddy_track_length*radar_dt; % conver to hours

figure('color','w')
subplot(231)
histogram(ed1)
xlabel('Eddy duration (hours)')
ylabel('Number of eddies')


subplot(232)
histogram(intensity)
xlabel('intensity (avg streamlines)')
ylabel('Number of eddies')
box off
grid on

subplot(233)
histogram(speed)
xlabel('speed (m/s)')
box off
grid on

subplot(234)
histogram(diameter)
xlabel('avg diameter (km)')
box off
grid on

subplot(235)
histogram(mindiameter)
xlabel('min diameter (km)')
box off
grid on

subplot(236)
histogram(maxdiameter)
xlabel('max diameter (km)')
box off
grid on

File: python/analyze_eddy_tracks.py
#!/usr/bin/env python3
"""Python port of analyze_eddy_tracks.m — plot the eddy tracks found by
eddy_tracking.py and compute track statistics.

Figure 1: track map (blue = CCW/cyclonic, red = CW/anticyclonic; green o =
start, black x = end) over coastline + bathymetry.
Figure 2: histograms of duration, intensity, speed, and diameters.
Both are saved to data/results_figs/.
"""

import os

import numpy as np
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
from scipy.io import loadmat

from utm_dc import geog2utm_nodisp
from etopo1_reader import read_etopo1

os.chdir(os.path.dirname(os.path.abspath(__file__)))
if not os.path.isdir('data') and os.path.isdir(os.path.join('..', 'data')):
    os.chdir('..')  # repo layout: python/ subfolder, data/ at the repo root

plt_individual = 0
highrescoast = 1
lw = 1
min_steps = 12   # minimum track length to plot
radar_dt = 0.5   # timestep in hours

# %% load
name_dir = 'data/results/'
name_pre = 'data2_'
t = loadmat(name_dir + name_pre + 'tracks.mat')
data = loadmat('data/data2.mat')
lons = data['lons']; lats = data['lats']

lon_center = [np.ravel(c) for c in t['lon_center'][0]]
lat_center = [np.ravel(c) for c in t['lat_center'][0]]
omega = [np.ravel(c) for c in t['omega'][0]]
num_streams = [np.ravel(c) for c in t['num_streams'][0]]
eig1 = [np.ravel(c) for c in t['eig1'][0]]
eig2 = [np.ravel(c) for c in t['eig2'][0]]
direction = [np.ravel(c) for c in t['direction'][0]]
Time = [np.ravel(c) for c in t['Time'][0]]
timegap = np.ravel(t['timegap'])
eddy_track_time_param = float(np.ravel(t['eddy_track_time_param'])[0])

# %% setup
lon_min = float(np.floor(10 * np.nanmin(lons)) / 10)
lon_max = float(np.ceil(10 * np.nanmax(lons)) / 10)
lat_min = float(np.floor(10 * np.nanmin(lats)) / 10)
lat_max = float(np.ceil(10 * np.nanmax(lats)) / 10)

# %% per-track statistics
ntracks = len(timegap)
print('max gap between eddy identification in timesteps = %g' % eddy_track_time_param)
print('number of tracks = %d' % ntracks)

eddy_track_length = np.array([len(d) for d in direction])
intensity = np.array([np.mean(s) for s in num_streams])
diameter = np.array([np.mean(e1 + e2) for e1, e2 in zip(eig1, eig2)])
maxdiameter = np.array([np.max(e1 + e2) for e1, e2 in zip(eig1, eig2)])
mindiameter = np.array([np.min(e1 + e2) for e1, e2 in zip(eig1, eig2)])
dist = np.full(ntracks, np.nan)
speed = np.full(ntracks, np.nan)
for i in range(ntracks):
    if eddy_track_length[i] > 1:
        x, y, _ = geog2utm_nodisp(lon_center[i], lat_center[i])
        d1 = np.hypot(x[-1] - x[0], y[-1] - y[0]) * 1000
        dt = (Time[i][-1] - Time[i][0]) * 24 * 3600
        dist[i] = d1
        speed[i] = d1 / dt

print('max eddy track length in timesteps = %d' % np.max(eddy_track_length))

# %% track map
order = np.argsort(-eddy_track_length)  # descending

fig1 = plt.figure(figsize=(11, 9), facecolor='w')
ax = fig1.add_subplot(111)
try:
    elev, elon, elat = read_etopo1(lon_min - 0.05, lon_max + 0.05,
                                   lat_min - 0.05, lat_max + 0.05)
    levels = [-2500, -1000, -500, -250, -100, -50, -25, -10]
    ax.contour(elon, elat, elev, levels=levels, colors=[(0.5, 0.5, 0.5)],
               linewidths=0.5, linestyles='-', zorder=1)
    land = np.ma.masked_less(elev, 0)
    ax.contourf(elon, elat, land, levels=[0, elev.max() + 1],
                colors=[(0.8, 0.8, 0.8)], zorder=2)
    ax.contour(elon, elat, elev, levels=[0], colors='k', linewidths=0.7, zorder=3)
except FileNotFoundError as e:
    print('warning: %s (skipping bathymetry/coastline)' % e)

counter = 0
title_set = False
for i in order:
    if eddy_track_length[i] < min_steps:
        continue
    counter += 1
    w1 = np.nanmean(omega[i])
    w2 = w1 * 3600 * 24 / 360  # rotations per day
    lonp = lon_center[i]
    latp = lat_center[i]
    color = 'b' if w1 > 0 else 'r'
    ax.plot(lonp, latp, '-' + color + '.', linewidth=lw, markersize=5, zorder=4)
    ax.plot(lonp[0], latp[0], 'o', linewidth=2, markerfacecolor='none',
            markeredgecolor=(0, 0.8, 0), markeredgewidth=2, zorder=5)
    ax.plot(lonp[-1], latp[-1], 'kx', markersize=9, markeredgewidth=2, zorder=5)
    if not title_set:
        ax.set_title('longest track = %d timesteps  ;   %.1f rot/day'
                     % (eddy_track_length[i], round(w2, 1)))
        title_set = True

ax.set_xlim(lon_min, lon_max)
ax.set_ylim(lat_min, lat_max)
ax.set_aspect(1 / np.cos(np.deg2rad((lat_min + lat_max) / 2)))
ax.set_xlabel('Longitude')
ax.set_ylabel('Latitude')
fig1.savefig('data/results_figs/data2_analysis_tracks.png', dpi=200,
             bbox_inches='tight')
plt.close(fig1)
print('%d tracks with >= %d timesteps plotted' % (counter, min_steps))

# %% histogram plots
fig2 = plt.figure(figsize=(13, 7), facecolor='w')
panels = [
    (eddy_track_length * radar_dt, 'Eddy duration (hours)'),
    (intensity, 'intensity (avg streamlines)'),
    (speed[~np.isnan(speed)], 'speed (m/s)'),
    (diameter, 'avg diameter (km)'),
    (mindiameter, 'min diameter (km)'),
    (maxdiameter, 'max diameter (km)'),
]
for ip, (vals, lab) in enumerate(panels, start=1):
    axh = fig2.add_subplot(2, 3, ip)
    axh.hist(vals, bins=10, edgecolor='k', linewidth=0.5)
    axh.set_xlabel(lab)
    if ip == 1:
        axh.set_ylabel('Number of eddies')
    axh.grid(True, alpha=0.4)
fig2.tight_layout()
fig2.savefig('data/results_figs/data2_analysis_hist.png', dpi=200,
             bbox_inches='tight')
plt.close(fig2)
print('figures saved to data/results_figs/')
